cnxarchive/views.py

```diff
@@ -1024,3 +1274,42 @@ def recent(request):
                                            ident_hash=module['uuid'])
     request.response.content_type = 'application/rss+xml'
     return {"latest_modules": latest_modules}
+
+
+@view_config(route_name='oai', request_method='GET',
+             renderer='templates/oai_verbs.xml')
+def oai(request):
+    request.response.headers = {'Content-Type': 'text/xml; charset=UTF-8'}
+
+    return_vars = {}
+    return_vars['dateTime'] = html_rss_date(datetime.now().time())
+    return_vars['baseURL'] = request.path_url
+    return_vars['query_request'] = []
+    for var in request.GET:
+        return_vars['query_request'].append({'name': var,
+                                             'val': request.GET[var]})
+
+    verb = request.GET.get('verb')
+    parse_results = _parseOaiArguments(verb, request)
+    return_vars['verb'] = verb
+    if parse_results != "No errors":
+        return_vars['error'] = parse_results
+        return return_vars
+
+    # Conditional statemtns for each verb
+    if verb == 'Identify':
+        _oaiIdentify(return_vars, request)
+    elif verb == 'ListMetadataFormats':
+        _oaiListMetadataFormats(return_vars, request)
+    if verb == 'ListSets':
+        _oaiListSets(return_vars, request)
+    if verb == 'ListIdentifiers':
+        _oaiListIdentifiers(return_vars, request)
+    if verb == 'SearchRecords':
+        _oaiSearchRecords(return_vars, request)
+    elif verb == 'GetRecord':
+        _oaiGetRecords(return_vars, request)
+    elif verb == 'ListRecords':
+        _oaiListRecords(return_vars, request)
```

Comment:
This is begging for a Dispatcher pattern, where the oai.py would provide a dictionary of the actual execution functions (which traditionally are named do_<action>), keyed on the verb. Also, returning values in an argument (particularly the first argument) is not very pythonic. I'd rather see the parsing and error catch, then the dispatch (which does the actual queries against the db), returning the result dict, which is used to update the results dict:
something like: `result_vars.update(oaiQueries(verb)(request))`